syntax = "proto3";

package mpointers;

// Service definition for the Memory Manager
service MemoryManagerService {
    // Create a new memory block
    rpc Create (CreateRequest) returns (CreateResponse);
    
    // Set a value in memory
    rpc Set (SetRequest) returns (SetResponse);
    
    // Get a value from memory
    rpc Get (GetRequest) returns (GetResponse);
    
    // Increase reference count
    rpc IncreaseRef (IncreaseRefRequest) returns (IncreaseRefResponse);
    
    // Decrease reference count
    rpc DecreaseRef (DecreaseRefRequest) returns (DecreaseRefResponse);
}

// Request/Response messages for Create
message CreateRequest {
    uint32 size = 1;
    string type = 2;
}

message CreateResponse {
    uint32 id = 1;
    bool success = 2;
    string error_message = 3;
}

// Request/Response messages for Set
message SetRequest {
    uint32 id = 1;
    bytes value = 2;
}

message SetResponse {
    bool success = 1;
    string error_message = 2;
}

// Request/Response messages for Get
message GetRequest {
    uint32 id = 1;
}

message GetResponse {
    bytes value = 1;
    bool success = 2;
    string error_message = 3;
}

// Request/Response messages for IncreaseRef
message IncreaseRefRequest {
    uint32 id = 1;
}

message IncreaseRefResponse {
    bool success = 1;
    string error_message = 2;
}

// Request/Response messages for DecreaseRef
message DecreaseRefRequest {
    uint32 id = 1;
}

message DecreaseRefResponse {
    bool success = 1;
    string error_message = 2;
} 